### DISCUSSION TITLE
OpenTelemetry Not Adding Listener For ActivitySource

### DISCUSSION THREAD

**Discussion by CoalChris** [open-telemetry/opentelemetry-dotnet#3669] [Q&A]
**Why is OpenTelemetry not adding a listener for my `ActivitySource` in a hosted service on generic host?**

I have some configurations to pass into OpenTelemetry set-up, so I depend on a class `MyClass`. This means registering that class, and then configuring a callback action for `TracerProviderBuilder` which uses that class. However, when I later create an `ActivitySource` and start an `Activity` with `ActivitySource.StartActivity(...)` then it returns null because no listeners were attached to the source. By debugging and inspecting the `ActivitySource`, I could see the list of `s_activeSources` which included the `OpenTelemetry.HttpRequest` `ActivitySource` created by `.AddHttpClientInstrumentation()` so it's odd that no listener was added for that either.

**Note: This is being run in a unit test and failing, whereas it is working in a long-running worker service program. Not sure if unit tests are too short-lived or are influenced by test environment e.g. Visual Studio**

Below is a minimal version of my scenario:
```
Host.CreateDefaultBuilder()
.ConfigureServices((context, services) =>
{
    services
    .AddSingleton<MyClass>()
    .AddOpenTelemetryTracing(hostingBuilder =>
        hostingBuilder
        .Configure(
            (sp, builder) =>
            {
                var myClass = sp.GetRequiredService<MyClass>();
                
                // Do something with myClass on builder
                
                builder
                .AddAspNetInstrumentation()
                .AddHttpClientInstrumentation()
                .AddConsoleExporter()
                .AddSource("TestService*");
            }))
    .AddHostedService<TestService>();
})
.Build();
```

`TestService.cs`
```
internal class TestService : BackgroundService
{
    private static ActivitySource testActivitySource = new ActivitySource("TestService");

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        await Task.Run(() =>
        {
            using (var activity = testActivitySource.StartActivity("TestService", ActivityKind.Server))
            {
                Console.WriteLine("I'm inside the activity using clause!");

                activity?.SetTag("foo", 1);

                Console.WriteLine($"activity.Tags: {activity.Tags}"); // This throws NullReferenceException since activity is null due to no listeners
            }

            Console.WriteLine("I'm outside the activity using clause!");
        });
    }
}
```

Package versions:
```
<package id="OpenTelemetry" version="1.3.0" />
<package id="OpenTelemetry.Api" version="1.3.0" />
<package id="OpenTelemetry.Exporter.Console" version="1.3.0" />
<package id="OpenTelemetry.Extensions.Hosting" version="1.0.0-rc9.5" />
<package id="OpenTelemetry.Instrumentation.AspNet" version="1.0.0-rc9.5" />
<package id="OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule" version="1.0.0-rc9.5" />
<package id="OpenTelemetry.Instrumentation.Http" version="1.0.0-rc9.5" />
```

I've also tried moving the `.AddConsoleExporter()` line out of the `.Configure` method and simply called straight away by `hostingBuilder` but still no listener attached to any `ActivitySource`.

**Comment by kccarter76** [comment]
I am having a similar issue where a listener is not registered with the ActivitySource.

`
public static CustomOpenTelemetryBuilder WithTracing(this CustomOpenTelemetryBuilder builder, Action<CustomTelemetryOptions>? telemetry = null,  Action<OtlpExporterOptions>? exporter = null)
{
    builder.Options.EnableTracing = true;

    telemetry ??= (CustomTelemetryOptions options) =>
    {
        options.AddContentTypeToTrace("*/json");
    };

    builder.Configure(telemetry);

    if (exporter != null)
    {
        builder.Options.ConfigureSection(exporter);

        builder.Services.Configure<OtlpExporterOptions>(builder.Options.SetSection);
    }

    builder.Services.AddOpenTelemetry()
            .WithTracing(tracer =>
            {
                if (builder.Options.Environment == Environments.DevUnstable)
                {
                    tracer.AddConsoleExporter();
                }

                if (builder.Options.HasSection<OtlpExporterOptions>())
                {
                    tracer.AddOtlpExporter(builder.Options.SetSection);
                }
                else
                {
                    tracer.AddCustomExporter(builder.Options.SetSection);
                }

                tracer.AddSource(TraceHttpActivityAccessor.Source.Name);
            });

    builder.Services.Configure<CustomTelemetryOptions>(builder.SetConfiguration); // hoping this will sync the configured objects in the IOC.

    return builder;
}
`

**Reply by cijothomas** [reply]
https://cloud-native.slack.com/archives/C01N3BC2W7Q/p1668557737052439 Same as this? The background service maybe starting earlier than SDK is initialized.

**Reply by kccarter76** [reply]
I do not have access to this work place on slack

**Reply by kccarter76** [reply]
I was actually able to determine why, apparently open telemetry if the recording flag is not set for parent traces it will not create an activity object. in that circumstance might consider restarting the trace if that span of a distributed trace is to be recorded.
